### PR TITLE
Fix issues with WcfLogReceiverClient

### PR DESCRIPTION
**Pull request by MartinTherriault**
(see #784)
I restarted from the original code and added the method to the client interface. For the server interface, I inherited the original interface and added the new method. The target kept its parameter to indicate which method to use.

There is however one possibility in which the changes could break old code, it is if someone was inheriting directly the client interface, they would have to implement the new method for the code to compile.

I do not think we can do otherwise:
- If we use two client interfaces (V1 and V2) which are exclusives, we end up with the same issues as before because we must inherit from ClientBase< T >
- If we use two client interfaces which V2 inherit from V1, it is fine until you try to configure a V1 endpoint, which will fail (throws) since we used a ClientBase< V2 > and it does not find any ClientBase< V1 >

### CODE REVIEW

File: src/NLog/LogReceiverService/ILogReceiverServerV2.cs [hunk 1]
```diff
@@ -41,14 +41,14 @@ namespace NLog.LogReceiverService
     /// Service contract for Log Receiver server.
     /// </summary>
     [ServiceContract(Namespace = LogReceiverServiceConfig.WebServiceNamespace)]
-    public interface ILogReceiverOneWayServer
```
Comment (304NotModified) [inline src/NLog/LogReceiverService/ILogReceiverServerV2.cs]: We lost #if WCF_SUPPORTED

File: src/NLog/LogReceiverService/WcfLogReceiverClient.cs [hunk 2]
```diff
@@ -237,12 +265,22 @@ protected override ILogReceiverClient CreateChannel()
...
+            }
+            else {
+                return ((ILogReceiverClient)this).BeginProcessLogMessages(events, callback, asyncState);
+            }
         }
 
         private object[] OnEndProcessLogMessages(IAsyncResult result)
         {
-            ((ILogReceiverClient)this).EndProcessLogMessages(result);
+            if (UseOneWayCallsToServer) {
```
Comment (304NotModified) [inline src/NLog/LogReceiverService/WcfLogReceiverClient.cs]: We don't need to copy the `if (UseOneWayCallsToServer)` (code duplication). I advise a helper method who's returning a `ILogReceiverClient`

**Comment by 304NotModified** [comment]
I'm not sure if this is the best solution. I will also do a try, and then make a PR. Let see if we can fix all the issues.

**Comment by 304NotModified** [comment]
Hi, it tried also something. See #821. I tried also removing the duplicate code in the clients, but failed

**Comment by MartinTherriault** [comment]
Ok, so do you want me to do now?

**Comment by 304NotModified** [comment]
Maybe check the code and give your opinion? :)  which approach is the best? Would it work?

**Comment by MartinTherriault** [comment]
I'm going to try to give a neutral opinion on both approach, but keep in mind I'm biased toward the #803 one. I believe both would work but one would require a higher maintenance effort, hence why I favor #803.

**Facade (#821)**
Pros
 * Ensure the WCF contract in non breaking, since we do not modify the original one
 
Cons
 * `ClientBase` inheritance force to have code duplication
 * Since the facade itself (`WcfLogReceiverClient`) cannot inherit from client base, it will break old code where someone is using a `ClientBase` property we did not implement. Using the proxy instead would likely fix the issue, otherwise a new method/property would be needed on `WcfLogReceiverClient`. (#783)
 * A server would need two endpoints to support both method since it is different contracts
 
**WCF versionning (#803)**
Pros
 * Adding a method to the contract is non breaking and inheriting the server contract ensure a server will not need any changes unless he wants to support the new method
 * A server can support both methods with a single endpoint
 * Client class (`WcfLogReceiverClient`) inherit `ClientBase`
 
Cons
 * Will break old code if someone was inheriting `ILogReceiverClient`. Implementing the new methods in the old code would fix the issue.
 
 
In both case, to support the new one way method, a server will have to change code, be it inherit another interface, implement it, add or not a new endpoint, ...

**Comment by 304NotModified** [comment]
Thanks for the analysis! 

I don't use WCF much. But I think that most of the time people only would implement one of the server interfaces?

**Comment by MartinTherriault** [comment]
Yes most of the time.

I do not think having to implement both methods is a burden though. Keep in mind if you only want to use the old two way method, the server can use the old interface and do not have to implement the other method. Also both method have the same signature so they can call the same processing method. The difference (one way vs two way) is done behind the curtains by the WCF layer.

**Comment by kevindaub** [comment]
This might work for the code duplication and might combine both ideas.  This will also bring in the ClientBase.

```csharp
public abstract class WcfLogReceiverClientBase<TInterface> : ClientBase<TInterface>
        where TInterface : ILogReceiverClient
{
}

public sealed class WcfLogReceiverOneWayClient : WcfLogReceiverClientBase<ILogReceiverOneWayClient>, ILogReceiverOneWayClient
{
}
```

I'll take #821 and test out the above idea and I'll get something together this weekend.

**Comment by 304NotModified** [comment]
superseded by #874

**Pull request closed without merging**